BUG: Fix segfault in csv tokenizer

roberthdevries opened the pull request:
- [x] closes #28071
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

roberthdevries committed:
Fix segfault in csv tokenizer (issue #28071)

roberthdevries committed:
Add whatsnew entry

WillAyd commented on pandas/_libs/src/parser/tokenizer.c:
```diff
@@ -1189,8 +1189,13 @@ int parser_consume_rows(parser_t *self, size_t nrows) {
     word_deletions = self->line_start[nrows - 1] + self->line_fields[nrows - 1];
-    char_count = (self->word_starts[word_deletions - 1] +
-                  strlen(self->words[word_deletions - 1]) + 1);
+    if (word_deletions >= 1) {
```
Do you have a handle on why this is required in the first place? I think there is a more general problem with the statement management that this works around (maybe OK)

roberthdevries replied:
As I have said in the ticket, I am also not certain *why* this works (apart from the obvious fact that this guards against a negative index in the array).
I took the heuristic route and created a test and fiddled around with a couple of options (setting charcount to 0 and to 1).
As it seemed to work and did not break any other unit tests, I thought that it would be useful to submit as a PR.
However, I agree that it does not feel good not knowing exactly *why* this works.
Maybe I put too much trust in unit testing, in the sense that this may break other functionality without unit tests.

gfyoung replied:
@roberthdevries : Is there a test that you added which fails when `char_count = 0` ?

roberthdevries replied:
if `char_count` is 0, there will be an endless loop. So that option was abandoned :smile:

gfyoung replied:
With respect to the test you wrote, this loop would manifest for all values of `nrows`

Also, add a comment about `char_count` needing to be at least one.

roberthdevries replied:
The original ticket only mentions `nrows=1`. I cannot guarantee that the bug would trigger on other values, only the test would not progress further than the first one.
I added the other numbers of `nrows` to be sure that my fix did not introduce unwanted behavior when multiple rows were being read.

I did some more experiments and the value that makes some kind of sense to me is `self->stream_len`. If there are no `word_deletions`, the parser must consume something, and the maximum that can be skipped is the `stream_len`.
With this value all the tests in the `pandas/tests/io/parser` directory keep working.

gfyoung replied:
The fact you can make that modification and still have tests passing concerns me a little.  It is hard to tell whether or not there is coverage for the `else` case.

To confirm, if you set `char_count = 0` in the `else` block, do you get test failures?  If not, you may need to add a `sprintf` debug command to see when this case ever gets hit.

roberthdevries replied:
I tested this case, and it passes the test just fine. (Not sure why it failed in the past).
I am investigating this a bit more, stay tuned. But for now it does not seem to matter much which value I put in there, as long as it is smaller or equal to `stream_len`

roberthdevries replied:
The patch is changed to `char_count = 0`. As `word_deletions == 0`, and `char_count` is used to skip characters, it seems logical to me that if `word_deletions == 0` that `char_count` must be 0 as well.

gfyoung replied:
I have to say that it's still a little weird that we can't pinpoint an exact reason for setting the value of `char_count`, but at least this number I can understand.  Can you add a comment to explain that reasoning?

roberthdevries replied:
Added

WillAyd commented on pandas/tests/io/parser/test_textreader.py:
```diff
@@ -341,6 +341,16 @@ def test_empty_csv_input(self):
+            columns=list("ab"),
+        )
+        csv = "\nheader\n\na,b\n\n\n1,2\n\n3,4"
+        for nrows in range(1, 6):
```
Instead of doing this can you just parametrize on the values that matter?

roberthdevries replied:
Can you elaborate which values you would like to see parameterized?
I basically took the example in the ticket to reproduce the issue and checked that for all sensible values for `nrows` I get the expected outcome.
The reason for this is that this covers more cases that could cause errors like the one in the original ticket due to various edge cases.

gfyoung replied:
@roberthdevries : @WillAyd is asking that you use `pytest.mark.parameterize` over the `nrows` parameter.

roberthdevries replied:
fixed

WillAyd commented on pandas/tests/io/parser/test_textreader.py:
```diff
@@ -341,6 +341,16 @@ def test_empty_csv_input(self):
         assert isinstance(df, TextFileReader)
 
+    def test_blank_lines_between_header_and_data_rows(self):
+        ref = DataFrame(
```
Can you add the GH issue number as a comment?

roberthdevries replied:
done

WillAyd reviewed (changes requested):
Great thanks for the PR @roberthdevries . We don't get a ton of contributions to the C extensions so much appreciate what you are doing here

roberthdevries committed:
Add ticket number to test

roberthdevries committed:
Untabify

roberthdevries committed:
Make linter happy

jreback commented on pandas/tests/io/parser/test_textreader.py:
```diff
@@ -341,6 +341,17 @@ def test_empty_csv_input(self):
         df = read_csv(StringIO(), chunksize=20, header=None, names=["a", "b", "c"])
         assert isinstance(df, TextFileReader)
 
+    def test_blank_lines_between_header_and_data_rows(self):
```
this test needs to be in test_common.py and use the all_parsers fixture so it testsed.

roberthdevries replied:
done

roberthdevries committed:
nrows is now parametrized

pep8speaks commented:
Hello @roberthdevries! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/tests/io/parser/test_textreader.py`](https://github.com/pandas-dev/pandas/blob/5007c283606a614732cc18e8cf7313d960dfb212/pandas/tests/io/parser/test_textreader.py):

> [Line 344:46](https://github.com/pandas-dev/pandas/blob/5007c283606a614732cc18e8cf7313d960dfb212/pandas/tests/io/parser/test_textreader.py#L344): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

roberthdevries committed:
Test moved from test_textreader to test_common

roberthdevries committed:
Skip the remainder of stream if there are no word_deletions

roberthdevries committed:
Set char_count to 0 when word_deletions is 0

This looks like the most sensible value as the number of words that is deleted
is 0 therefore the char_count of characters to be skipped is also 0.

roberthdevries committed:
Add comment to explain the reasoning to set char_count to 0

jreback commented:
@roberthdevries pls merge master; ping on green.

roberthdevries commented:
ping @jreback everything is green

jreback commented:
thanks @roberthdevries
